### PR TITLE
Document `docs/link` shortcode

### PR DESCRIPTION
**Pull request by jdbaldry**
For context on the need, refer to [2023-04 Flexible linking -- Problem](https://docs.google.com/document/d/1F9aiHhyruW0d_zMoL5yjG37Qo-OZ--x4Ivx2Hdu1cYs/edit#heading=h.xxn36c2bhtyg) and [2023-04 Flexible linking -- Consensus](https://docs.google.com/document/d/1F9aiHhyruW0d_zMoL5yjG37Qo-OZ--x4Ivx2Hdu1cYs/edit#heading=h.ve4du26wqsib).

I'd like to use this documentation to drive the implementation.
The documentation should help a user use this shortcode with confidence, if anything is unclear or you believe it is overly complicated, please add comments to this PR and I will incorporate the feedback in both the documentation and the implementation.

Signed-off-by: Jack Baldry <jack.baldry@grafana.com>

### CODE REVIEW

File: docs/sources/writing-guide/shortcodes/index.md [hunk 1]
```diff
@@ -41,6 +41,34 @@ Kingston is the capital of Jamaica.
 {{%/* /admonition */%}}
 ```
 
+## `docs/link` shortcode
+
+The `docs/link` shortcode offers more flexible linking than the Hugo builtin `relref` shortcode.
+Each argument to the shortcode has the form `<PATH PREFIX> -> <REFERENCE>`.
+_`PATH PREFIX`_ is matched against the page's URL path.
```
Comment (osg-grafana) [inline docs/sources/writing-guide/shortcodes/index.md:48]:
```suggestion
Each argument to the shortcode has the form `<PATH PREFIX> -> <REFERENCE>`, where _`PATH PREFIX`_ is matched against the page's URL path.
```

File: docs/sources/writing-guide/shortcodes/index.md [hunk 2]
```diff
@@ -41,6 +41,34 @@ Kingston is the capital of Jamaica.
...
+If _`PATH PREFIX`_ matches, then _`REFERENCE`_ is used as the argument to Hugo's `relref` shortcode.
+
+_`REFERENCE`_ can include a version variable that is looked up in the page's front matter.
```
Comment (osg-grafana) [inline docs/sources/writing-guide/shortcodes/index.md:51]:
```suggestion
The `docs/link` shortcode offers more flexible linking than the Hugo builtin `relref` shortcode.
Each argument to the shortcode has the form `<PATH PREFIX> -> <REFERENCE>`:
* _`PATH PREFIX`_ matches against the page's URL path. If _`PATH PREFIX`_ matches, then _`REFERENCE`_ is used as the argument to Hugo's `relref` shortcode.
* _`REFERENCE`_ can include a version variable that is looked up in the page's front matter.
```

File: docs/sources/writing-guide/shortcodes/index.md [hunk 3]
```diff
@@ -41,6 +41,34 @@ Kingston is the capital of Jamaica.
...
+The version variable must be enclosed with `<` and `>` symbols, and the final word must be `VERSION`.
+For example, `<GRAFANA VERSION>`, or `<MIMIR VERSION>`.
+The version variable is lower cased and spaces are replaced with underscores (`_`) when looking up the variable in front matter.
+Using the preceding examples, `<GRAFANA VERSION>` becomes `grafana_version` and `<MIMIR VERSION>` becomes `mimir_version`.
+
+### Example
+
+The following example has behavior such that:
+
+- If the shortcode is used in the page `/docs/grafana/latest/dashboards/`, and the page does not have the front matter `grafana_version`, the shortcode returns a relref to `/docs/grafana/latest`.
```
Comment (osg-grafana) [inline docs/sources/writing-guide/shortcodes/index.md:62]: There is no `dashboards` after `/docs/grafana/latest/`?

Comment (jdbaldry) [reply docs/sources/writing-guide/shortcodes/index.md:62]: The documentation here is correct given the proceeding usage.

Because of the argument `"/docs/grafana/ -> /docs/grafana/<GRAFANA VERSION>"`, the link returned in pages that have the prefix `/docs/grafana/` is `/docs/grafana/<GRAFANA VERSION>` and because there is no front matter, `/docs/grafana/<GRAFANA VERSION>` becomes `/docs/grafana/latest`.

It seems I might need to reword the explanation for clarity or perhaps the syntax of the shortcode is confusing?

File: docs/sources/writing-guide/shortcodes/index.md [hunk 4]
```diff
@@ -41,6 +41,34 @@ Kingston is the capital of Jamaica.
...
+_`REFERENCE`_ can include a version variable that is looked up in the page's front matter.
+If the front matter variable is not found, then the shortcodes uses the version inferred from the page's URL path.
+The version variable must be enclosed with `<` and `>` symbols, and the final word must be `VERSION`.
+For example, `<GRAFANA VERSION>`, or `<MIMIR VERSION>`.
+The version variable is lower cased and spaces are replaced with underscores (`_`) when looking up the variable in front matter.
+Using the preceding examples, `<GRAFANA VERSION>` becomes `grafana_version` and `<MIMIR VERSION>` becomes `mimir_version`.
+
+### Example
+
+The following example has behavior such that:
```
[the remark this answers is not in the document]
Comment (jdbaldry) [reply docs/sources/writing-guide/shortcodes/index.md:60]: I think I need to be more clear that the behaviors described are the results of the specific example in `{{< docs/link "/docs/grafana/ -> /docs/grafana/<GRAFANA VERSION>" "/docs/grafana-cloud/ -> /docs/grafana-cloud" "/docs/mimir/ -> /docs/grafana/<GRAFANA VERSION>" >}}`.

Should the example come first and the explanation after?

File: docs/sources/writing-guide/shortcodes/index.md [hunk 5]
```diff
@@ -41,6 +41,34 @@ Kingston is the capital of Jamaica.
...
+
+- If the shortcode is used in the page `/docs/grafana/latest/dashboards/`, and the page does not have the front matter `grafana_version`, the shortcode returns a relref to `/docs/grafana/latest`.
+
+- If the shortcode is used in the page `/docs/grafana-cloud/introduction/`, the shortcode returns a relref to `/docs/grafana-cloud`.
+
+- If the shortcode is used in the page `/docs/mimir/latest/dashboards/`, and the page has the front matter `grafana_version: next`, the shortcode returns a relref to `/docs/grafana/next`.
```
Comment (osg-grafana) [inline docs/sources/writing-guide/shortcodes/index.md:66]:
```suggestion
- If the shortcode is used in the page `/docs/mimir/latest/dashboards/`, and the page has the front matter `mimir_version: next`, the shortcode returns a relref to `/docs/mimir/next`.
```
I think you mean `mimir` here.

File: docs/sources/writing-guide/shortcodes/index.md [hunk 6]
```diff
@@ -41,6 +41,34 @@ Kingston is the capital of Jamaica.
...
+
+```markdown
+{{</* docs/link "/docs/grafana/ -> /docs/grafana/<GRAFANA VERSION>" "/docs/grafana-cloud/ -> /docs/grafana-cloud" "/docs/mimir/ -> /docs/grafana/<GRAFANA VERSION>" */>}}
```
Comment (osg-grafana) [inline docs/sources/writing-guide/shortcodes/index.md:69]: Is this `GRAFANA` or `MIMIR` for this example?

File: docs/sources/writing-guide/shortcodes/index.md [hunk 7]
```diff
@@ -41,6 +41,34 @@ Kingston is the capital of Jamaica.
...
+The version variable is lower cased and spaces are replaced with underscores (`_`) when looking up the variable in front matter.
+Using the preceding examples, `<GRAFANA VERSION>` becomes `grafana_version` and `<MIMIR VERSION>` becomes `mimir_version`.
+
+### Example
+
+The following example has behavior such that:
+
+- If the shortcode is used in the page `/docs/grafana/latest/dashboards/`, and the page does not have the front matter `grafana_version`, the shortcode returns a relref to `/docs/grafana/latest`.
+
+- If the shortcode is used in the page `/docs/grafana-cloud/introduction/`, the shortcode returns a relref to `/docs/grafana-cloud`.
```
[the remark this answers is not in the document]
Comment (jdbaldry) [reply docs/sources/writing-guide/shortcodes/index.md:64]: The documentation here is correct given the proceeding usage.

Because of the argument `"/docs/grafana-cloud -> /docs/grafana-cloud`", the link returned in pages that have the prefix `/docs/grafana-cloud/` (part before ` -> `) is `/docs/grafana-cloud` (part after ` -> `).

File: docs/sources/writing-guide/shortcodes/index.md (hunk 5 above)
Comment (jdbaldry) [reply docs/sources/writing-guide/shortcodes/index.md:66]: The documentation here is correct given the proceeding usage.

Because of the argument `"/docs/mimir/ -> /docs/grafana/<GRAFANA VERSION>`", the link returned in pages that have the prefix `/docs/mimir/` (part before ` -> `) is `/docs/grafana/<GRAFANA VERSION>` (part after ` -> `), and because there is the front matter `grafana_version: next`, `/docs/grafana/<GRAFANA VERSION>` becomes `/docs/grafana/next`.

File: docs/sources/writing-guide/shortcodes/index.md (hunk 6 above)
Comment (osg-grafana) [reply docs/sources/writing-guide/shortcodes/index.md:69]: I find this quite confusing... I will try to say this in plain English based on my first impression:

"I would like to link to Grafana docs in general and then to a specific version. Then I might want to think about `grafana-cloud` too, and go to Grafana Cloud documentation and also Mimir documentation, which goes to the version of the Grafana docs that I just specified."

Comment (osg-grafana) [reply docs/sources/writing-guide/shortcodes/index.md:69]: Are there two things happening here:
* Link related to Grafana
* Link related to Grafana Cloud

Do we need two links here to break these down into individual instructions to be transformed?

**Review by osg-grafana — CHANGES_REQUESTED** [review]
Thank you for being so open to feedback, and I am happy to noodle over this with you to make it clearer.

File: docs/sources/writing-guide/shortcodes/index.md (hunk 6 above)
Comment (jdbaldry) [reply docs/sources/writing-guide/shortcodes/index.md:69]: Thank you for trying to understand this shortcode and explain your thoughts.
It's clear that its behavior is not obvious and I need to think how to improve that.

There are three things happening, each one is explained by the preceding list items.

This shortcode has three different results.
Which result is returned is dependent on matching the page it is used in against the _`PATH PREFIX`_ part of the parameters.

If a _`PATH PREFIX`_ matches the page, then the related _`REFERENCE`_ is returned. That reference might contain version variables that are interpolated.

File: docs/sources/writing-guide/shortcodes/index.md (hunk 2 above)
Comment (jdbaldry) [reply docs/sources/writing-guide/shortcodes/index.md:51]: Edited your suggestion to make it match `prettier` style.

File: docs/sources/writing-guide/shortcodes/index.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (osg-grafana) [reply docs/sources/writing-guide/shortcodes/index.md:61]: I think that this formatting (if possible) would make it more clear that there are three things happening here:

```
{{< docs/link
  "/docs/grafana/ -> 
   /docs/grafana/<GRAFANA VERSION>" 
  
  "/docs/grafana-cloud/ ->
   /docs/grafana-cloud"
  
  "/docs/mimir/ ->
   /docs/grafana/<GRAFANA VERSION>" >}}.
```

[the remark this answers is not in the document]
Comment (jdbaldry) [reply docs/sources/writing-guide/shortcodes/index.md:61]: ;; This buffer is for text that is not saved, and for Lisp evaluation.
;; To create a file, visit it with C-x C-f and enter text in its buffer.

Unfortunately, I don't think that can work especially when the shortcode is used in links that are contained within a sentence.

Perhaps we could use reference style links and different shortcode syntax to make it more clear:

```markdown
[link text][reference]
...

{{% docs/link %}}
[reference]: "/docs/grafana/ -> /docs/grafana/<GRAFANA VERSION>"
[reference]: "/docs/grafana-cloud/ -> /docs/grafana-cloud"
[reference]: "/docs/mimir/ -> /docs/grafana/<GRAFANA VERSION>"
{{% /docs/link %}}
```

After shortcode rendering, I would expect something like:

1. For `/docs/grafana/latest/dashboards/`
   ```markdown
   [link text][reference]
   ...
   
   [reference]: "/docs/grafana/latest"
   ```
1. For `/docs/grafana-cloud/introduction/`
   ```markdown
   [link text][reference]
   ...
   
   [reference]: "/docs/grafana-cloud"
   ```
1. For `/docs/mimir/latest/dashboards/`, with front matter `grafana_version: next`
   ```markdown
   [link text][reference]
   ...
   
   [reference]: "/docs/grafana/next"
   ```

The problem with the above is that we don't really use reference much throughout our repositories.

**Comment by jdbaldry** [comment]
Superseded by `docs/reference`
